File: 1669-merge-in-between-linked-lists/1669-merge-in-between-linked-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeInBetween(self, list1: ListNode, a: int, b: int, list2: ListNode) -> ListNode:
        start = list1

        while a > 1:
            start = start.next
            a -= 1
            b -= 1

        end = start.next
        b -= 1

        while b:
            end = end.next
            b -= 1
        
        start.next = list2
        curr = start.next

        while curr.next:
            curr = curr.next

        curr.next = end.next
        end.next = None

        return list1

File: 1669-merge-in-between-linked-lists/1669-merge-in-between-linked-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeInBetween(ListNode* list1, int a, int b, ListNode* list2) {
        ListNode* start = list1;
        ListNode* prevStart = nullptr;
        ListNode* end = list1;
        while (b > 0) {
            if (a > 0) {
                prevStart = start;
                start = start->next;
                --a;
            }
            end = end->next;
            --b;
        }
        if (prevStart != nullptr) prevStart->next = list2;
        ListNode* list2Ptr = prevStart->next;
        while (list2Ptr->next != nullptr) list2Ptr = list2Ptr->next;
        list2Ptr->next = end->next;
        list2 = nullptr;
        return list1;
    }
};